File: app/shoe.mjs
import { CARD_VALUES, SHOE_BASE, MODIFIERS_TOTAL, ACTIONS_TOTAL, MODIFIER_TYPES, ACTION_TYPES } from './deck.mjs';

export function createEmptySeen() {
  const seen = { other: 0, modifierTypes: {}, actionTypes: {} };
  for (const v of CARD_VALUES) {
    seen[v] = { regular: 0, special: 0 };
  }
  for (const id of Object.keys(MODIFIER_TYPES)) seen.modifierTypes[id] = 0;
  for (const id of Object.keys(ACTION_TYPES)) seen.actionTypes[id] = 0;
  return seen;
}

export function logCardSeen(seen, value, kind) {
  const next = {
    other: seen.other,
    modifierTypes: { ...seen.modifierTypes },
    actionTypes: { ...seen.actionTypes },
  };
  for (const v of CARD_VALUES) next[v] = { ...seen[v] };

  if (value === 'other') {
    const max = MODIFIERS_TOTAL + ACTIONS_TOTAL;
    if (seen.other + 1 > max) {
      throw new Error(`Cannot log another "other" card: all ${max} already seen`);
    }
    next.other = seen.other + 1;
    return next;
  }

  const max = SHOE_BASE[value][kind];
  if (seen[value][kind] + 1 > max) {
    throw new Error(`Cannot log another ${value}/${kind}: all ${max} already seen`);
  }
  next[value][kind] = seen[value][kind] + 1;
  return next;
}

export function remainingCount(seen, value, kind) {
  return SHOE_BASE[value][kind] - seen[value][kind];
}

export function totalUndrawn(seen) {
  let seenTotal = seen.other;
  for (const v of CARD_VALUES) {
    seenTotal += seen[v].regular + seen[v].special;
  }
  const totalCards =
    CARD_VALUES.reduce((sum, v) => sum + SHOE_BASE[v].regular + SHOE_BASE[v].special, 0) +
    MODIFIERS_TOTAL +
    ACTIONS_TOTAL;
  return totalCards - seenTotal;
}

function sumTypeCounts(typeCounts) {
  return Object.values(typeCounts).reduce((sum, n) => sum + n, 0);
}

export function logSpecialCardSeen(seen, category, id) {
  const types = category === 'modifier' ? MODIFIER_TYPES : ACTION_TYPES;
  const max = types[id].max;
  const currentTypeCounts = category === 'modifier' ? seen.modifierTypes : seen.actionTypes;
  if (currentTypeCounts[id] + 1 > max) {
    throw new Error(`Cannot log another ${id}: all ${max} already seen`);
  }

  const next = {
    other: seen.other,
    modifierTypes: { ...seen.modifierTypes },
    actionTypes: { ...seen.actionTypes },
  };
  for (const v of CARD_VALUES) next[v] = { ...seen[v] };

  if (category === 'modifier') {
    next.modifierTypes[id] = seen.modifierTypes[id] + 1;
  } else {
    next.actionTypes[id] = seen.actionTypes[id] + 1;
  }
  next.other = sumTypeCounts(next.modifierTypes) + sumTypeCounts(next.actionTypes);
  return next;
}

File: desktop/src/flip7_desktop/shoe.py
from .deck import CARD_VALUES, SHOE_BASE, MODIFIER_TYPES, ACTION_TYPES, TOTAL_CARDS


def create_empty_seen() -> dict:
    seen = {"other": 0, "modifier_types": {}, "action_types": {}}
    for v in CARD_VALUES:
        seen[v] = {"regular": 0, "special": 0}
    for type_id in MODIFIER_TYPES:
        seen["modifier_types"][type_id] = 0
    for type_id in ACTION_TYPES:
        seen["action_types"][type_id] = 0
    return seen


def log_card_seen(seen: dict, value: int, kind: str) -> dict:
    next_seen = {
        "other": seen["other"],
        "modifier_types": dict(seen["modifier_types"]),
        "action_types": dict(seen["action_types"]),
    }
    for v in CARD_VALUES:
        next_seen[v] = dict(seen[v])

    max_count = SHOE_BASE[value][kind]
    if seen[value][kind] + 1 > max_count:
        raise ValueError(f"Cannot log another {value}/{kind}: all {max_count} already seen")
    next_seen[value][kind] = seen[value][kind] + 1
    return next_seen


def remaining_count(seen: dict, value: int, kind: str) -> int:
    return SHOE_BASE[value][kind] - seen[value][kind]


def total_undrawn(seen: dict) -> int:
    seen_total = seen["other"]
    for v in CARD_VALUES:
        seen_total += seen[v]["regular"] + seen[v]["special"]
    return TOTAL_CARDS - seen_total


def _sum_type_counts(type_counts: dict) -> int:
    return sum(type_counts.values())


def log_special_card_seen(seen: dict, category: str, type_id: str) -> dict:
    types = MODIFIER_TYPES if category == "modifier" else ACTION_TYPES
    max_count = types[type_id]["max"]
    key = "modifier_types" if category == "modifier" else "action_types"
    current_count = seen[key][type_id]
    if current_count + 1 > max_count:
        raise ValueError(f"Cannot log another {type_id}: all {max_count} already seen")

    next_seen = {
        "other": seen["other"],
        "modifier_types": dict(seen["modifier_types"]),
        "action_types": dict(seen["action_types"]),
    }
    for v in CARD_VALUES:
        next_seen[v] = dict(seen[v])

    next_seen[key][type_id] = current_count + 1
    next_seen["other"] = _sum_type_counts(next_seen["modifier_types"]) + _sum_type_counts(next_seen["action_types"])
    return next_seen
